core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -63,80 +66,136 @@ public class SharePartitionManager {
+        } finally {
+            lock.unlock();
+        }
+    }
 
-            replicaManager.fetchMessages(
-                fetchParams,
-                CollectionConverters.asScala(
-                    topicPartitionData.entrySet().stream().map(entry -> new Tuple2<>(entry.getKey(), entry.getValue())).collect(
-                        Collectors.toList())
-                ),
-                QuotaFactory.UnboundedQuota$.MODULE$,
-                responsePartitionData -> {
-                    List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
-                        responsePartitionData);
-                    Map<TopicIdPartition, ShareFetchResponseData.PartitionData> result = new HashMap<>();
-                    responseData.forEach(data -> {
-                        TopicIdPartition topicIdPartition = data._1;
-                        FetchPartitionData fetchPartitionData = data._2;
-
-                        partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition))
-                            .acquire(memberId, fetchPartitionData).whenComplete((acquiredRecords, throwable) -> {
-                                ShareFetchResponseData.PartitionData partitionData = new ShareFetchResponseData.PartitionData()
-                                    .setPartitionIndex(topicIdPartition.partition());
-
-                                if (throwable != null) {
-                                    partitionData.setErrorCode(Errors.forException(throwable).code());
-                                } else {
-                                    // Maybe check if no records are acquired and we want to retry replica
-                                    // manager fetch. Depends on the share partition manager implementation,
-                                    // if we want parallel requests for the same share partition or not.
-                                    partitionData
-                                        .setPartitionIndex(topicIdPartition.partition())
-                                        .setRecords(fetchPartitionData.records)
-                                        .setErrorCode(fetchPartitionData.error.code())
-                                        .setAcquiredRecords(acquiredRecords)
-                                        .setAcknowledgeErrorCode(Errors.NONE.code());
-                                }
-                                result.put(topicIdPartition, partitionData);
-                            });
+    public boolean isFetchQueueEmpty() {
+        lock.lock();
+        try {
+            return fetchQueue.isEmpty();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public ShareFetchPartitionData pollFetchQueue() {
+        lock.lock();
+        try {
+            return fetchQueue.poll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void fetchResults() {
```

Comment:
Seems the method is called everytime in `fetchMessages` and the method `fetchResults` is blocking which means you cannot return `future` from `fetchMessages` unless the `fetchResults` method completes, and which is not right.
1. The method `fetchResults` should be renamed to `maybeProcessFetchQueue`.
2. `maybeProcessFetchQueue` should have a CAS check on AtomicBoolean regarding if the fetch is already going on.
3. Which means if the queue is already being processed then anything which is added in the queue will eventually be processed and we can return our future.
4. When we come out of fetchQueue loop then we must set the CAS check to false so the next `fetchMessges` request which adds anything further in queue can trigger the queue processing again.

Reply:
Also processing the method in while loop seems not right, if you receive multiple requests together then the first request which triggered the fetch processing of queue will be blocked, though future of that request will be completed but it could never be returned to the caller unless queue finishes.

So we either do it like this:
1. Check post getting response from `replicaManager` call which does 2 things:
    a. Sets the active fetch queue sets (AtomicBoolean) to false. 
    b. Calls the `maybeProcessFetchQueue` method if queue is not empty, which triggers the queue processing again.
This approach will move the queue sequentially as like we cannot have parallel calls for different topic partitions on replica manager, which is an issue as well but may be good to start with.

2. Execute the processing of the queue on the different thread which ll make sure the `fetchMessages` requests are not blocked and we can have parallel calls going to replica manager as well.

Reply:
thanks @apoorvmittal10 for explaining it on the standup. I'll go ahead with the implementation as you suggested.